File: pe-core/src/main/java/org/saar/core/model/FutureMesh.java
package org.saar.core.model;

import java.util.concurrent.CompletableFuture;

public class FutureMesh implements Mesh {

    private FutureMeshHelper helper;

    private FutureMesh(FutureMeshHelper helper) {
        this.helper = helper;
    }

    public static FutureMesh create(CompletableFuture<? extends Mesh> task) {
        return new FutureMesh(FutureMeshHelper.create(task));
    }

    public static FutureMesh unloaded(CompletableFuture<? extends UnloadedMesh> task) {
        return new FutureMesh(FutureMeshHelper.unloaded(task));
    }

    @Override
    public void draw() {
        this.helper = this.helper.next();
        this.helper.draw();
    }

    @Override
    public void delete() {
        this.helper = this.helper.next();
        this.helper.delete();
    }
}

File: pe-core/src/main/java/org/saar/core/mesh/async/FutureMesh.kt
package org.saar.core.mesh.async

import org.saar.core.mesh.Mesh
import org.saar.core.mesh.UnloadedMesh
import java.util.concurrent.CompletableFuture

class FutureMesh private constructor(private var helper: FutureMeshHelper) : Mesh {

    override fun draw() {
        this.helper = this.helper.next()
        this.helper.draw()
    }

    override fun delete() {
        this.helper = this.helper.next()
        this.helper.delete()
    }

    companion object {
        fun create(task: CompletableFuture<out Mesh>): FutureMesh {
            val helper = FutureMeshHelper.create(task)
            return FutureMesh(helper)
        }

        fun unloaded(task: CompletableFuture<out UnloadedMesh>): FutureMesh {
            val helper = FutureMeshHelper.unloaded(task)
            return FutureMesh(helper)
        }
    }
}
